### PR TITLE
[NFC][CONTRIBUTING] Extend contribution guidelines

### PR DESCRIPTION
**Pull request by enirolf**
This follows the proposal and discussion during the ROOT team meeting on 22/01/2024 (https://indico.cern.ch/event/1368009/). The goal for these new guidelines is to keep the git history as organized as possible, as well as have something to point (new) contributors to.

A few related points/questions remain open:
- [x] Do we want to keep both `[NFC]` and `[skip-ci]` tags (what is the difference, if any)?
- [ ] Do we want to we enable automatic labeling of PRs, e.g. through https://github.com/actions/labeler?
- [ ] How are the scopes for commit summaries defined?

### CODE REVIEW

**Review by couet — APPROVED** [review]

File: CONTRIBUTING.md [hunk 1]
```diff
@@ -1,35 +1,77 @@
-
-How to Contribute Code to ROOT
-==============================
+# How to Contribute Code to ROOT
```
Comment (dpiparo) [inline CONTRIBUTING.md:1]: We should at some point make clear that these are guidelines, as it was done during the ROOT meeting.

File: CONTRIBUTING.md [hunk 2]
```diff
@@ -1,35 +1,77 @@
...
+2. **Each commit does not contain more than one independent change.**
+This allows us to revert changes when needed, without affecting anything else.
+
+> [!TIP]
+> During a code review, it may be useful to make smaller commits to track intermediate changes, and rebase after the PR
+> is approved to ensure the above points are met and to reduce clutter.
+
+### Your Commit Message
+
+The commit summary (i.e. the first line of the commit message) should be preceded by the part of ROOT that is affected
```
Comment (dpiparo) [inline CONTRIBUTING.md:40]: We should be ready to list the parts of ROOT and write that if the collaborator does find anything appropriate, we are here to help.

File: CONTRIBUTING.md [hunk 3]
```diff
@@ -1,35 +1,77 @@
...
+by your commit, in square brackets. See the [commit log](https://github.com/root-project/root/commits/master/) for
+examples. If you're not sure which scope you change belongs to, don't hesitate to ask! The summary itself should not
+exceed 50 characters, be meaningful (i.e., it describes the change) and should be written in the
```
Comment (dpiparo) [inline CONTRIBUTING.md:43]: We should be explicit whether this is 50 modulo the commit scope or including it.

File: CONTRIBUTING.md [hunk 4]
```diff
@@ -1,35 +1,77 @@
...
+The commit message should be wrapped at 72 characters.
+
+## Your Pull Request
+
+> [!NOTE]
+> For the mechanics on how to create pull requests, please visit
+> [this page](https://root.cern/for_developers/creating_pr).
+
+The title of your PR follows the same principle as the commit summary. If your PR only involves one commit, you can
+reuse this summary. For non-functional changes (e.g. to the documentation) that don't require a ROOT build, use `[NFC]`
```
Comment (dpiparo) [inline CONTRIBUTING.md:59]: Should we make 2 cases here? If the change is non functional, we want to label it as [NFC] and skip the ci. However, we might also have a change which changes functionality but for some reason skip temporarily the ci with [skip-ci], for example while exposing changes at a very early stage.

File: CONTRIBUTING.md [hunk 5]
```diff
@@ -1,35 +1,77 @@
...
+change diff, respectively).
+The commit message should be wrapped at 72 characters.
+
+## Your Pull Request
+
+> [!NOTE]
+> For the mechanics on how to create pull requests, please visit
+> [this page](https://root.cern/for_developers/creating_pr).
+
+The title of your PR follows the same principle as the commit summary. If your PR only involves one commit, you can
```
Comment (dpiparo) [inline CONTRIBUTING.md:58]: Does the 50 chars limit hold also for PR titles?

File: CONTRIBUTING.md [hunk 6]
```diff
@@ -1,35 +1,77 @@
...
 Of course you will be duly credited: for sizable contributions your name will appear in the
-[CREDITS](https://raw.githubusercontent.com/root-project/root/master/README/CREDITS){:target="_blank"}
+[CREDITS](https://raw.githubusercontent.com/root-project/root/master/README/CREDITS)
 file shipped with every binary and source distribution.
 The copyright transfer helps us with effectively defending the project in case of litigation.
 
-:warning: We require PRs to cleanly apply to master without a merge commit, i.e. through "fast-forward".
-Please follow the [coding conventions](https://root.cern.ch/coding-conventions),
-as this is a simple item for reviewers to otherwise get stuck on.
+## Your Commit
```
Comment (dpiparo) [inline CONTRIBUTING.md:25]: I am not sure about this one, but the commits of some projects have in their description something like "Fixes #<12345>". Would this be useful for us?

File: CONTRIBUTING.md [hunk 7]
```diff
@@ -43,17 +85,15 @@ Suppose for your PR you create a branch on `root.git`.
 Our CI infrastructure automatically picks up a branch with the same name in your fork of `roottest.git`
```
Comment (dpiparo) [inline CONTRIBUTING.md:85]: Is this really clear? Would something like "The modified or added test can be in root or in roottest. Given the branch name in the ROOT fork used to make the PR,  our CI infrastructure automatically picks up the changes defined in the roottest PR which is made from an identical branch name, if it exists."

File: CONTRIBUTING.md [hunk 8]
```diff
@@ -43,17 +85,15 @@ Suppose for your PR you create a branch on `root.git`.
...
+## Continuous Integration
 
 To prevent bad surprises and make a better first impression, we
 strongly encourage new developers to [run the tests](https://root.cern/for_developers/run_the_tests/)
 _before_ submitting a pull request.
 
 ROOT has automated CI tests :cop: that are used for pull requests:
 - *Build and test*: a [Jenkins-based CI workflow](https://github.com/phsft-bot/build-configuration/blob/master/README.md)
-    tests PRs automatically; a project member might need to initiate this build.
+    as well as a GitHub Actions CI workflow tests PRs automatically; a project member might need to initiate this build.
```
Comment (dpiparo) [inline CONTRIBUTING.md:96]: "Only a project member is allowed to initiate this build."

**Review by dpiparo — COMMENTED** [review]
Thanks for this work, it makes easier for collaborators to contribute. I just flagged my review as "Comment", waiting also the opinion of others to converge on the largest consensus possible.

File: CONTRIBUTING.md (hunk 4 above)
Comment (bellenot) [reply CONTRIBUTING.md:59]: @dpiparo OK, fair enough, I didn't think about this use-case...

**Comment by dpiparo** [comment]
Now that I see this issue (https://its.cern.ch/jira/projects/ROOT/issues/ROOT-8915) can we perhaps also provide a template for the commit messages, and close the issue?

**Comment by enirolf** [comment]
> Now that I see this issue (https://its.cern.ch/jira/projects/ROOT/issues/ROOT-8915) can we perhaps also provide a template for the commit messages, and close the issue?

Yes, this seems like a nice addition. The points raised by Philippe in the comments are still valid, we can either only provide it as an opt-in template that people themselves have to add locally as a commit template, or we have to enforce it somehow in the CI. I'm not sure the latter is desirable as it would turn the guidelines into rules. I don't know if this would be possible, but what could be nice is a 'commit format checker' in the CI that doesn't immediately turn red if it doesn't succeed but rather points the contributor to the guidelines.

File: CONTRIBUTING.md (hunk 5 above)
Comment (enirolf) [reply CONTRIBUTING.md:65]: Good question, I don't really have a strong opinion on this but I think in principle we could be a bit more lenient here (maybe use the 72 chars from the commit message) since they are usually only viewed in the web interface.

File: CONTRIBUTING.md (hunk 6 above)
Comment (enirolf) [reply CONTRIBUTING.md:30]: I (sort of) mention this on line 65. In my opinion this should be part of the PR description/title rather than the commit since it's tightly coupled to the GitHub issue/PR machinery.

File: CONTRIBUTING.md (hunk 2 above)
Comment (enirolf) [reply CONTRIBUTING.md]: Do we want to add this list to this document, or only keep it internally?

File: CONTRIBUTING.md [hunk 9]
```diff
@@ -1,59 +1,109 @@
...
-----------------------
+Whatever it is: this requires you to add a new test, or to extend an existing test. Depending on the size and complexity
+of this test, it exists either in the `test/` subdirectory of each part of ROOT (see for instance
+[`tree/dataframe/test`](https://github.com/root-project/root/tree/master/tree/dataframe/test)), or in
+[roottest](https://github.com/root-project/roottest.git). Tests in `test/` subdirectories are unittests, mostly based on
+[Google Test](https://github.com/google/googletest) and easily extended. Tests in
+[roottest](https://github.com/root-project/roottest.git) are more involved (e.g., tests requiring custom dictionaries or
+data files). When you create a branch in the main ROOT repository (i.e., this repository) and add a test to `roottest`,
+make sure to do this under the same branch name (and open a PR for it). Our CI infrastructure automatically picks up the
+changes defined in the `roottest` PR based on this branchname, and uses that for testing your PR here.
```
Comment (martamaja10) [inline CONTRIBUTING.md:95]: I like this addition about the tests! 

Maybe it's a corner case*  - but I stumbled upon this recently - that if you only add new tests in the roottest directory, in order to test if they don't break the new CI, you need to open a PR in the root repo (with some pseudo change, like an extra space in some description), using the same branch name as the one in the roottest, so that both CI's are tested. 

*and maybe more internal for ourselves in the team until roottest also tests for the new CI, if that's in the plans?

Comment (pcanal) [reply CONTRIBUTING.md:95]: > *and maybe more internal for ourselves in the team until roottest also tests for the new CI, if that's in the plans?

A PR adding proper roottest sync to the new CI is almost done/merged.

Comment (vepadulano) [reply CONTRIBUTING.md:95]: While true, I believe that's unrelated to what Marta is mentioning here. Even with the sync of roottest when pushing PRs to root for the new CI, we still need to enable the new CI for those PRs that are pushed to roottest only, without a root sibling

File: CONTRIBUTING.md [hunk 10]
```diff
@@ -1,59 +1,109 @@
...
-==============================
+# How to Contribute Code to ROOT
 
 Thank you for your interest in contributing to ROOT!  We strongly welcome and appreciate such contributions!
 This short guide tries to make contributing as quick and painless as possible.
 
-Your Pull Request
------------------------
+> [!NOTE]
+> These guidelines should be applicable to most contributes. At the same time, these are not 'one-size-fits-all' rules,
```
Comment (vepadulano) [inline CONTRIBUTING.md:7]:
```suggestion
> These guidelines should be applicable to most contributions. At the same time, these are not 'one-size-fits-all' rules,
```

File: CONTRIBUTING.md [hunk 11]
```diff
@@ -1,59 +1,109 @@
...
+
+> [!TIP]
+> During a code review, it may be useful to make smaller commits to track intermediate changes, and rebase after the PR
+> is approved to ensure the above points are met and to reduce clutter.
+
+### Your Commit Message
+
+The commit summary (i.e. the first line of the commit message) should be preceded by the a tag indicating the scope of
+ROOT that is affected by your commit, in square brackets. Most tags are self-describing (e.g., `[tree]` indicates a
+change to TTree, `[RF]` indicates a change to RooFit). If you are unsure about which scope tags to use, we are happy to
```
Comment (vepadulano) [inline CONTRIBUTING.md:47]: It would be nicer to have a list of these tags (I don't think we ever created one). We could address this in a separate PR. Also, it would be nice to decide on the capitalization, e.g. `tree` vs `TREE`, `RF` vs `rf` etc.

File: CONTRIBUTING.md [hunk 12]
```diff
@@ -1,59 +1,109 @@
...
+point you in the right direction! See also the [commit log](https://github.com/root-project/root/commits/master/) for
+examples. The summary itself should not exceed 50 characters (excluding the scope tag), be meaningful (i.e., it
+describes the change) and should be written in the
+[present imperative mood](https://git.kernel.org/pub/scm/git/git.git/tree/Documentation/SubmittingPatches?id=HEAD#n239)
+(e.g. `Add this awesome feature` instead of `Adds this awesome feature` or `Added this awesome feature`).
+
+The commit message that follow the summary can be used to provide more context to the change.
```
Comment (vepadulano) [inline CONTRIBUTING.md:54]:
```suggestion
The commit message that follows the summary can be used to provide more context to the change.
```

**Review by vepadulano — APPROVED** [review]
Many thanks for this! Just a couple of minor comments

File: CONTRIBUTING.md (hunk 9 above)
Comment (pcanal) [reply CONTRIBUTING.md:95]: Indeed.  Once the source repository CI is fully function we should implement it for `roottest` (and it should not be too hard because it is probably similar to the source repository's)

**Review by martamaja10 — APPROVED** [review]

**Review by jblomer — APPROVED** [review]
